### PR TITLE
chore(ci): create release with semantic-release

### PR DESCRIPTION
**Pull request by soonum**
This will perform automatic release on GitHub. It will append a CHANGELOG file that will be versioned. It will create a new version tag and create a release entry on GitHub along with all the changes in the same format as CHANGELOG.

### CODE REVIEW

File: .releaserc [hunk 1]
```diff
@@ -0,0 +1,35 @@
+{
+  "tagFormat": "v${version}",
```
Comment (IceTDrinker) [inline .releaserc:2]: I don't think we put the v in the tag when we push a tag

File: scripts/prepare_release.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+#
+
+CARGO_FILE="tfhe/Cargo.toml"
+DOCS_DIR="tfhe/docs"
+DOC_FILES="getting_started/installation.md
+core_crypto/tutorial.md
+high_level_api/serialization.md
+high_level_api/tutorial.md"
+
+NEWTAG=$1
```
Comment (IceTDrinker) [inline scripts/prepare_release.sh:13]: new version rather ?

File: .github/workflows/create_release.yml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+          additional_packages: |
+            [
+              '@semantic-release/github',
+              '@semantic-release/git',
+              '@semantic-release/changelog',
+              '@semantic-release/exec',
+            ]
+          dry_run: ${{ github.event.client_payload.dry_run }}
+        env:
+          GITHUB_TOKEN: ${{ secrets.CONCRETE_ACTIONS_TOKEN }}
```
Comment (IceTDrinker) [inline .github/workflows/create_release.yml:33]: no crates.io push ?

File: .releaserc [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  "verifyConditions": ["@semantic-release/github"],
+  "plugins": [
+    "@semantic-release/commit-analyzer",
+    "@semantic-release/release-notes-generator",
+    [
+      "@semantic-release/changelog",
+      {
+        "changelogFile": "CHANGELOG.md",
```
[the remark this answers is not in the document]
Comment (soonum) [reply .releaserc:10]: Look at the root of Slab repository.
Note that the change log in the Release section can still be modified afterward to add some context for example.

File: scripts/prepare_release.sh [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+# Update Cargo.toml package version
+sed -i "s|^version = \"[[:digit:]]*\.[[:digit:]]*\.[[:digit:]]*\"|version = \"${NEWTAG}\"|" $CARGO_FILE
+
+# Update package version in documentation
+for file in $DOC_FILES; do
+  sed -i "s|^\(tfhe = { version = \)\"[[:digit:]]*\.[[:digit:]]*\.[[:digit:]]*\"|\1\"${NEWTAG}\"|" "$DOCS_DIR/$file"
+done
```
[the remark this answers is not in the document]
Comment (soonum) [reply scripts/prepare_release.sh:21]: As long as a line starts with `tfhe = { version = x.y.z`, no I don't think so.
Maybe we could remove the carret to force the regex to match the beginning of the line, wydt ?

[the remark this answers is not in the document]
Comment (IceTDrinker) [reply scripts/prepare_release.sh:21]: ah yeah, we could remove the caret

**Comment by IceTDrinker** [comment]
does this create commits for the changelog e.g. ?

File: .releaserc (hunk 4 above)
[the remark this answers is not in the document]
Comment (soonum) [reply .releaserc:10]: It's has been already seen with him :wink:

**Comment by soonum** [comment]
> does this create commits for the changelog e.g. ?

It creates a release commit containing all the changes (version upgrade, changelog updates, ...) and the version tag is associated to this commit.

**Comment by IceTDrinker** [comment]
> > does this create commits for the changelog e.g. ?
> 
> It creates a release commit containing all the changes (version upgrade, changelog updates, ...) and the version tag is associated to this commit.

Does it mean it needs to pass the CI ? 🤔

File: .github/workflows/make_release.yml [hunk 6]
```diff
@@ -3,16 +3,62 @@ name: Make release
...
+      - name: Build web package
+        run: |
+          make build_web_js_api
+
+      - name: Publish web package
+        uses: JS-DevTools/npm-publish@0f451a94170d1699fd50710966d48fb26194d939
+        with:
+            token: ${{ secrets.NPM_TOKEN }}
+            package: tfhe/pkg/package.json
+            dry-run: ${{ inputs.dry_run }}
```
Comment (IceTDrinker) [inline .github/workflows/make_release.yml:52]: I think some metadata needs to be checked with @immortal-tofu and be sure we are pushing something that corresponds to the npm package with the right name ?

File: .github/workflows/make_release.yml [hunk 7]
```diff
@@ -3,16 +3,62 @@ name: Make release
...
+
+      - name: Build Node package
+        run: |
+          make build_node_js_api
```
Comment (immortal-tofu) [inline .github/workflows/make_release.yml:56]: Does build_node_js_api clear the directory from the previous build?

File: .github/workflows/make_release.yml (hunk 6 above)
Comment (immortal-tofu) [reply .github/workflows/make_release.yml:52]: It is correct, we discuss it already with @soonum

File: .github/workflows/make_release.yml (hunk 7 above)
Comment (soonum) [reply .github/workflows/make_release.yml:56]: I can remove `pkg` directory prior to that to ensure this is a fresh package.

Comment (immortal-tofu) [reply .github/workflows/make_release.yml:63]: AFAIK, you build web package in same directory just before. That's why I suggested it.

**Review by IceTDrinker — COMMENTED** [review]
Small thing but otherwise I think we are there

**Pull request merged**